internal/engine/container_build_and_deployer.go

```diff
@@ -6,33 +6,93 @@ import (
 	"github.com/windmilleng/tilt/internal/model"
 )
 
 var _ BuildAndDeployer = &ContainerBuildAndDeployer{}
 
 type ContainerBuildAndDeployer struct {
-	cu  *build.ContainerUpdater
-	env k8s.Env
+	cu        *build.ContainerUpdater
+	env       k8s.Env
+	k8sClient k8s.Client
 
 	// containerBD can't do initial build, so will call out to ibd for that.
 	// May also fall back to ibd for certain error cases.
 	ibd ImageBuildAndDeployer
+
+	skipContainer bool
+}
+
+func DefaultSkipContainer() bool {
+	return true
 }
 
-// TODO(maia): wire this up
-func NewContainerBuildAndDeployer(cu *build.ContainerUpdater, env k8s.Env, ibd ImageBuildAndDeployer) *ContainerBuildAndDeployer {
+func NewContainerBuildAndDeployer(cu *build.ContainerUpdater, env k8s.Env, kCli k8s.Client, ibd ImageBuildAndDeployer, skipContainer bool) *ContainerBuildAndDeployer {
 	return &ContainerBuildAndDeployer{
-		cu:  cu,
-		env: env,
-		ibd: ibd,
+		cu:            cu,
+		env:           env,
+		k8sClient:     kCli,
+		ibd:           ibd,
+		skipContainer: skipContainer,
 	}
 }
 
 func (cbd *ContainerBuildAndDeployer) BuildAndDeploy(ctx context.Context, service model.Service, state BuildState) (BuildResult, error) {
 	span, ctx := opentracing.StartSpanFromContext(ctx, "daemon-ContainerBuildAndDeployer-BuildAndDeploy")
 	defer span.Finish()
+	if cbd.skipContainer {
+		return cbd.ibd.BuildAndDeploy(ctx, service, state)
+	}
+
+	// TODO(maia): proper output for this stuff
+
+	// TODO(maia): put service.Validate() upstream if we're gonna want to call it regardless
+	// of implementation of BuildAndDeploy?
+	err := service.Validate()
+	if err != nil {
+		return BuildResult{}, err
+	}
+
+	// ContainerBuildAndDeployer doesn't support initial build; call out to the ImageBuildAndDeployer
+	if state.IsEmpty() {
+		return cbd.ibd.BuildAndDeploy(ctx, service, state)
+	}
+	// Service has already been deployed; try to update in the running container
+
+	// If we don't know the pod that we just deployed to/container we just deployed, get it
+	var cID k8s.ContainerID
+	if state.LastResult.Container.String() == "" {
+		pID, err := cbd.k8sClient.PodWithImage(ctx, state.LastResult.Image)
+		if err != nil {
+			logger.Get(ctx).Infof("Unable to find pod, falling back to image deploy: %s", err.Error())
+			return cbd.ibd.BuildAndDeploy(ctx, service, state)
+		}
+		logger.Get(ctx).Infof("Deploying to pod: %s", pID)
+		// get containerID from pID (see container_updater.go --> containerIdForPod)
+		cID, err = cbd.cu.ContainerIDForPod(ctx, pID)
+		if err != nil {
+			logger.Get(ctx).Infof("Unable to find container, falling back to image deploy: %s", err.Error())
+			return cbd.ibd.BuildAndDeploy(ctx, service, state)
+		}
+		logger.Get(ctx).Infof("Got container ID for pod: %s", cID)
+	} else {
+		cID = state.LastResult.Container
+	}
+	// once have cID -- can call cbd.cu.UpdateContainer(...)
+	cf, err := build.FilesToPathMappings(state.FilesChanged(), service.Mounts)
+	if err != nil {
+		return BuildResult{}, err
```

Comment:
TODO: in some cases we may want to fall back to image build if the container build errors...?